main.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+const core = require('@actions/core');
+const github = require('@actions/github');
+const exec = require('@actions/exec');
+import { chdir } from 'process';
+const glob = require('@actions/glob');
+
+(async () => {
+  try {
+    const credentials = core.getInput('credentials');
+    const charm_path = core.getInput('charm-path');
+    const charmcraft_channel = core.getInput('charmcraft-channel');
+
+    chdir(charm_path);
+
+    await exec.exec('sudo', [
+      'snap',
+      'install',
+      'charmcraft',
+      '--classic',
+      '--channel',
+      charmcraft_channel,
+    ]);
+
+    await exec.exec('charmcraft', ['pack', '--destructive-mode', '-q']);
+
+    const globber = await glob.create('./*.charm');
+    const paths = await globber.glob();
+
+    await Promise.all(
+      paths.map((path) =>
+        exec.exec(
+          'charmcraft',
+          ['upload', '--release=latest/edge', path],
```

Comment:
We should expose release as an argument as well (with default of `latest/edge`).  We don't have any plans of using it that way, but I'm sure someone will

Reply:
oh shoot and the optional `branch` as well, that we'd need.  Not sure if it is better to specify like `latest/edge/somebranch` or have args for all three separately?

Reply:
IMO we should wait until there's an actual use case for it. I say this because even if someone wants to push to stable from github actions, that should be a separate job that's manually triggered by a person. It would also not be the `charmcraft upload` command again, it would be `charmcraft release`.

Reply:
As far as `branch` goes, I think this github action should do some automatic git parsing and push a branch automatically if possible. Some examples:

 - PR opened from `my-cool-bugfix` to `master` or `main`
   - This action publishes to the `latest/edge/my-cool-bugfix` charmhub branch
 - PR opened from `my-cool-bugfix` to `track/1.0`
   - This action publishes to the `1.0/edge/my-cool-bugfix` charmhub branch
 - PR merged into `master` or `main`
   - This action publishes to the `latest/edge` charmhub channel
 - PR merged into `track/1.0`
   - This action publishes to the `1.0/edge` charmhub channel

This has the advantage of being easily controlled by the `on` section of a workflow. For example, if you wanted to manage multiple tracks but didn't want to publish PRs, you could have this `on` section:

```yaml
on:
  push:
    branches:
      - main
      - track/**
```

Or if you wanted only some PRs to be published, you could do this:

```yaml
on:
  push:
    branches:
      - main
      - track/**
      - pr/**
```

And then only PRs with a git branch name starting with `pr/` would be published to a charmhub branch.

Reply:
+1 on the git parsing for branches and +1 on the release action probably being a separate action that would be triggered manually.

For the initial implementation `latest/edge` should be enough.

Reply:
I like the idea for automation a lot, that would address almost all cases really well.  

For cases where we wanted to be explicit, it would still be nice to directly call out where something goes (exactly specify track, branch, etc).  Maybe we have a basic action for uploading charms to an exact spot, and a second wrapper that uses the first action but adds additional logic to do the git parsing?

Reply:
There might be use cases that make sense for more explicit handling. I think we should wait until we run into a situation that calls for it before adding any functionality, though.There's a good chance we'll get things not quite right if we try to anticipate it now without a concrete use case.